File: internal/modules/universe/price_validator.go
package universe

import (
	"fmt"
	"math"
	"time"

	"github.com/rs/zerolog"
)

const (
	// Validation thresholds
	maxPriceMultiplier    = 10.0    // Price > 10x average is abnormal
	minPriceMultiplier    = 0.1     // Price < 0.1x average is abnormal
	maxPriceChangePercent = 1000.0  // >1000% change is a spike
	minPriceChangePercent = -90.0   // <-90% change is a crash
	absolutePriceMax      = 10000.0 // Absolute maximum price (EUR/USD)
	absolutePriceMin      = 0.01    // Absolute minimum price
	contextWindowDays     = 30      // Use last 30 days for context
)

// InterpolationLog records when a price was interpolated
type InterpolationLog struct {
	Date              string
	OriginalClose     float64
	InterpolatedClose float64
	Method            string // "linear", "forward_fill", "backward_fill"
	Reason            string
}

// PriceValidator validates and interpolates abnormal prices
type PriceValidator struct {
	log zerolog.Logger
}

// NewPriceValidator creates a new price validator
func NewPriceValidator(log zerolog.Logger) *PriceValidator {
	return &PriceValidator{
		log: log.With().Str("component", "price_validator").Logger(),
	}
}

// ValidatePrice checks if a price is valid
// Returns (isValid, reason)
func (v *PriceValidator) ValidatePrice(price DailyPrice, context []DailyPrice) (bool, string) {
	// 1. OHLC consistency checks (always applied, no context needed)
	if price.High < price.Low {
		return false, "high_below_low"
	}
	if price.High < price.Open {
		return false, "high_below_open"
	}
	if price.High < price.Close {
		return false, "high_below_close"
	}
	if price.Low > price.Open {
		return false, "low_above_open"
	}
	if price.Low > price.Close {
		return false, "low_above_close"
	}

	// 2. Percentage change checks (requires context)
	if len(context) > 0 {
		// Check day-over-day change FIRST (takes priority over average checks)
		prevClose := context[0].Close
		if prevClose > 0 {
			changePercent := ((price.Close - prevClose) / prevClose) * 100.0
			if changePercent > maxPriceChangePercent {
				return false, "spike_detected"
			}
			if changePercent < minPriceChangePercent {
				return false, "crash_detected"
			}
		}

		// Calculate average of recent prices (last 30 or all if <30)
		contextSize := len(context)
		if contextSize > contextWindowDays {
			contextSize = contextWindowDays
		}
		recentContext := context[:contextSize]

		var sum float64
		for _, p := range recentContext {
			sum += p.Close
		}
		avgPrice := sum / float64(len(recentContext))

		// Check if price is too high or too low relative to average
		if price.Close > avgPrice*maxPriceMultiplier {
			return false, "price_too_high"
		}
		if price.Close < avgPrice*minPriceMultiplier {
			return false, "price_too_low"
		}
	}

	// 3. Absolute bounds (fallback when no context)
	if len(context) == 0 {
		if price.Close > absolutePriceMax {
			return false, "absolute_bound_exceeded"
		}
		if price.Close < absolutePriceMin {
			return false, "absolute_bound_below_minimum"
		}
	}

	return true, ""
}

// InterpolatePrice interpolates an abnormal price using surrounding valid prices
// Returns (interpolatedPrice, method, error)
func (v *PriceValidator) InterpolatePrice(price DailyPrice, before []DailyPrice, after []DailyPrice) (DailyPrice, string, error) {
	interpolated := price // Start with original (preserves date, volume)

	// Parse dates for calculation
	priceDate, err := time.Parse("2006-01-02", price.Date)
	if err != nil {
		return interpolated, "", fmt.Errorf("failed to parse price date: %w", err)
	}

	// Linear interpolation (both before and after available)
	if len(before) > 0 && len(after) > 0 {
		beforePrice := before[0]
		afterPrice := after[0]

		beforeDate, err := time.Parse("2006-01-02", beforePrice.Date)
		if err != nil {
			return interpolated, "", fmt.Errorf("failed to parse before date: %w", err)
		}
		afterDate, err := time.Parse("2006-01-02", afterPrice.Date)
		if err != nil {
			return interpolated, "", fmt.Errorf("failed to parse after date: %w", err)
		}

		daysBetween := priceDate.Sub(beforeDate).Hours() / 24.0
		totalDays := afterDate.Sub(beforeDate).Hours() / 24.0

		if totalDays <= 0 {
			return interpolated, "", fmt.Errorf("invalid date range: totalDays <= 0")
		}

		// Interpolate Close
		interpolated.Close = beforePrice.Close + (afterPrice.Close-beforePrice.Close)*(daysBetween/totalDays)

		// Calculate ratios for Open/High/Low
		beforeOpenRatio := beforePrice.Open / beforePrice.Close
		afterOpenRatio := afterPrice.Open / afterPrice.Close
		openRatio := (beforeOpenRatio + afterOpenRatio) / 2.0

		beforeHighRatio := beforePrice.High / beforePrice.Close
		afterHighRatio := afterPrice.High / afterPrice.Close
		highRatio := (beforeHighRatio + afterHighRatio) / 2.0

		beforeLowRatio := beforePrice.Low / beforePrice.Close
		afterLowRatio := afterPrice.Low / afterPrice.Close
		lowRatio := (beforeLowRatio + afterLowRatio) / 2.0

		// Apply ratios to interpolated Close
		interpolated.Open = interpolated.Close * openRatio
		interpolated.High = interpolated.Close * highRatio
		interpolated.Low = interpolated.Close * lowRatio

		// Ensure OHLC consistency
		if interpolated.High < interpolated.Close {
			interpolated.High = interpolated.Close
		}
		if interpolated.Low > interpolated.Close {
			interpolated.Low = interpolated.Close
		}
		if interpolated.High < interpolated.Open {
			interpolated.High = interpolated.Open
		}
		if interpolated.Low > interpolated.Open {
			interpolated.Low = interpolated.Open
		}

		return interpolated, "linear", nil
	}

	// Forward fill (only before available)
	if len(before) > 0 {
		beforePrice := before[0]
		interpolated.Close = beforePrice.Close
		interpolated.Open = beforePrice.Open
		interpolated.High = beforePrice.High
		interpolated.Low = beforePrice.Low
		return interpolated, "forward_fill", nil
	}

	// Backward fill (only after available)
	if len(after) > 0 {
		afterPrice := after[0]
		interpolated.Close = afterPrice.Close
		interpolated.Open = afterPrice.Open
		interpolated.High = afterPrice.High
		interpolated.Low = afterPrice.Low
		return interpolated, "backward_fill", nil
	}

	// No context available - ensure OHLC consistency at minimum
	ensureOHLCConsistency(&interpolated)
	return interpolated, "no_interpolation", nil
}

// ValidateAndInterpolate validates all prices and interpolates abnormal ones
// Returns (validatedPrices, interpolationLogs, error)
func (v *PriceValidator) ValidateAndInterpolate(prices []DailyPrice, context []DailyPrice) ([]DailyPrice, []InterpolationLog, error) {
	if len(prices) == 0 {
		return prices, []InterpolationLog{}, nil
	}

	result := make([]DailyPrice, 0, len(prices))
	logs := []InterpolationLog{}

	for i, price := range prices {
		valid, reason := v.ValidatePrice(price, context)
		if valid {
			// Price is valid, keep as-is
			result = append(result, price)
			continue
		}

		// Price is abnormal, need to interpolate
		// Find before and after prices
		var before []DailyPrice
		var after []DailyPrice

		// Look for before price (previous valid price in prices array or result)
		// First check already processed valid prices in result
		for j := len(result) - 1; j >= 0; j-- {
			// Check if this result price is before the current price (by date)
			resultDate, err1 := time.Parse("2006-01-02", result[j].Date)
			currentDate, err2 := time.Parse("2006-01-02", price.Date)
			if err1 == nil && err2 == nil && resultDate.Before(currentDate) {
				before = []DailyPrice{result[j]}
				break
			}
		}
		// If not found in result, check previous prices in the prices array
		if len(before) == 0 {
			for j := i - 1; j >= 0; j-- {
				prevValid, _ := v.ValidatePrice(prices[j], context)
				if prevValid {
					before = []DailyPrice{prices[j]}
					break
				}
			}
		}
		// If still not found, find the most recent context price that's before current price
		if len(before) == 0 && len(context) > 0 {
			priceDate, err := time.Parse("2006-01-02", price.Date)
			if err == nil {
				for _, ctxPrice := range context {
					ctxDate, err := time.Parse("2006-01-02", ctxPrice.Date)
					if err == nil && ctxDate.Before(priceDate) {
						before = []DailyPrice{ctxPrice}
						break
					}
				}
			}
		}

		// Look for after price (next valid price in prices array)
		for j := i + 1; j < len(prices); j++ {
			// Check if next price is valid
			nextValid, _ := v.ValidatePrice(prices[j], context)
			if nextValid {
				after = []DailyPrice{prices[j]}
				break
			}
		}
		// If not found in prices, try to find a future date in context
		if len(after) == 0 && len(context) > 0 {
			priceDate, err := time.Parse("2006-01-02", price.Date)
			if err == nil {
				for _, ctxPrice := range context {
					ctxDate, err := time.Parse("2006-01-02", ctxPrice.Date)
					if err == nil && ctxDate.After(priceDate) {
						after = []DailyPrice{ctxPrice}
						break
					}
				}
			}
		}

		// Interpolate
		interpolated, method, err := v.InterpolatePrice(price, before, after)
		if err != nil {
			v.log.Error().
				Err(err).
				Str("date", price.Date).
				Msg("Failed to interpolate price, using original")
			// Use original price if interpolation fails
			result = append(result, price)
			continue
		}

		// Log interpolation
		logs = append(logs, InterpolationLog{
			Date:              price.Date,
			OriginalClose:     price.Close,
			InterpolatedClose: interpolated.Close,
			Method:            method,
			Reason:            reason,
		})

		v.log.Warn().
			Str("date", price.Date).
			Float64("original_close", price.Close).
			Float64("interpolated_close", interpolated.Close).
			Str("method", method).
			Str("reason", reason).
			Msg("Interpolated abnormal price")

		result = append(result, interpolated)
	}

	return result, logs, nil
}

// Helper function to ensure OHLC consistency
func ensureOHLCConsistency(price *DailyPrice) {
	// Ensure High >= all
	price.High = math.Max(price.High, price.Open)
	price.High = math.Max(price.High, price.Close)

	// Ensure Low <= all
	price.Low = math.Min(price.Low, price.Open)
	price.Low = math.Min(price.Low, price.Close)

	// Ensure High >= Low
	if price.High < price.Low {
		price.High = price.Low
	}
}

File: sentinel/price_validator.py
"""
Price Validator - Validates and interpolates abnormal price data.

Ported from sentinel2's internal/repositories/price_validator.go.

The validator:
1. Detects spikes (>1000% change) and crashes (<-90% change)
2. Checks if prices are >10x or <0.1x the rolling average
3. Validates OHLC consistency (High >= Low, etc.)
4. Interpolates invalid values using linear interpolation
5. Checks live prices for trade-blocking anomalies
"""

import logging
from dataclasses import dataclass
from typing import Optional, Tuple

logger = logging.getLogger(__name__)


# Validation thresholds
MAX_PRICE_MULTIPLIER = 10.0  # Price > 10x average is abnormal
MIN_PRICE_MULTIPLIER = 0.1  # Price < 0.1x average is abnormal
MAX_PRICE_CHANGE_PCT = 1000.0  # >1000% change is a spike
MIN_PRICE_CHANGE_PCT = -90.0  # <-90% change is a crash
CONTEXT_WINDOW_DAYS = 30  # Use last 30 days for context


@dataclass
class OHLCValidation:
    """Tracks validity of each OHLC component."""

    open_valid: bool = True
    high_valid: bool = True
    low_valid: bool = True
    close_valid: bool = True
    reason: str = ""

    def needs_full_interpolation(self) -> bool:
        """Returns True when Close is invalid (Close is the anchor)."""
        return not self.close_valid

    def needs_interpolation(self) -> bool:
        """Returns True when any component is invalid."""
        return not (self.open_valid and self.high_valid and self.low_valid and self.close_valid)

    def all_valid(self) -> bool:
        """Returns True when all components are valid."""
        return self.open_valid and self.high_valid and self.low_valid and self.close_valid


class PriceValidator:
    """Validates and interpolates abnormal prices."""

    def validate_price(self, price: dict, previous_price: Optional[dict], context: list[dict]) -> OHLCValidation:
        """
        Validate each OHLC component independently.
        Returns OHLCValidation with per-component validity status.
        """
        result = OHLCValidation()
        close = price.get("close", 0)

        # Zero/negative Close indicates missing or invalid data
        if close <= 0:
            return OHLCValidation(
                open_valid=False, high_valid=False, low_valid=False, close_valid=False, reason="close_zero_or_negative"
            )

        high = price.get("high", close)
        low = price.get("low", close)
        open_price = price.get("open", close)

        # 1. OHLC consistency checks
        if high < low:
            result.high_valid = False
            result.low_valid = False
            result.reason = "high_below_low"
        if high < open_price:
            result.high_valid = False
            if not result.reason:
                result.reason = "high_below_open"
        if high < close:
            result.high_valid = False
            if not result.reason:
                result.reason = "high_below_close"
        if low > open_price:
            result.low_valid = False
            if not result.reason:
                result.reason = "low_above_open"
        if low > close:
            result.low_valid = False
            if not result.reason:
                result.reason = "low_above_close"

        # 2. Day-over-day change detection (spike/crash affects all components)
        if previous_price and previous_price.get("close", 0) > 0:
            prev_close = previous_price["close"]
            change_pct = ((close - prev_close) / prev_close) * 100.0

            if change_pct > MAX_PRICE_CHANGE_PCT:
                return OHLCValidation(
                    open_valid=False, high_valid=False, low_valid=False, close_valid=False, reason="spike_detected"
                )
            if change_pct < MIN_PRICE_CHANGE_PCT:
                return OHLCValidation(
                    open_valid=False, high_valid=False, low_valid=False, close_valid=False, reason="crash_detected"
                )

        # 3. Average-based validation (requires context)
        if context and result.close_valid:
            context_size = min(len(context), CONTEXT_WINDOW_DAYS)
            avg_price = sum(p["close"] for p in context[:context_size]) / context_size

            if close > avg_price * MAX_PRICE_MULTIPLIER:
                return OHLCValidation(
                    open_valid=False, high_valid=False, low_valid=False, close_valid=False, reason="price_too_high"
                )
            if close < avg_price * MIN_PRICE_MULTIPLIER:
                return OHLCValidation(
                    open_valid=False, high_valid=False, low_valid=False, close_valid=False, reason="price_too_low"
                )

        return result

    def interpolate_price(
        self, price: dict, validation: OHLCValidation, before: list[dict], after: list[dict]
    ) -> tuple[dict, str]:
        """
        Interpolate invalid components of a price.
        Returns (interpolated_price, method).
        """
        interpolated = price.copy()

        # If Close is invalid, we need full interpolation
        if not validation.close_valid:
            return self._interpolate_full(price, before, after)

        # Selective interpolation - only fix invalid components
        method = "selective"

        if not validation.high_valid:
            ratio = self._get_typical_high_close_ratio(before, after)
            interpolated["high"] = interpolated["close"] * ratio

        if not validation.low_valid:
            ratio = self._get_typical_low_close_ratio(before, after)
            interpolated["low"] = interpolated["close"] * ratio

        if not validation.open_valid:
            if before:
                interpolated["open"] = before[0]["close"]
            elif after:
                interpolated["open"] = after[0].get("open", after[0]["close"])
            else:
                interpolated["open"] = interpolated["close"]

        self._ensure_ohlc_consistency(interpolated)
        return interpolated, method

    def _interpolate_full(self, price: dict, before: list[dict], after: list[dict]) -> tuple[dict, str]:
        """Perform full interpolation when Close is invalid."""
        interpolated = price.copy()

        # Linear interpolation if both before and after available
        if before and after:
            before_close = before[0]["close"]
            after_close = after[0]["close"]
            # Simple midpoint interpolation
            interpolated["close"] = (before_close + after_close) / 2.0

            # Use ratios for other components
            open_ratio = (
                before[0].get("open", before_close) / before_close + after[0].get("open", after_close) / after_close
            ) / 2.0
            high_ratio = (
                before[0].get("high", before_close) / before_close + after[0].get("high", after_close) / after_close
            ) / 2.0
            low_ratio = (
                before[0].get("low", before_close) / before_close + after[0].get("low", after_close) / after_close
            ) / 2.0

            interpolated["open"] = interpolated["close"] * open_ratio
            interpolated["high"] = interpolated["close"] * high_ratio
            interpolated["low"] = interpolated["close"] * low_ratio

            self._ensure_ohlc_consistency(interpolated)
            return interpolated, "linear"

        # Forward fill
        if before:
            interpolated["close"] = before[0]["close"]
            interpolated["open"] = before[0].get("open", before[0]["close"])
            interpolated["high"] = before[0].get("high", before[0]["close"])
            interpolated["low"] = before[0].get("low", before[0]["close"])
            return interpolated, "forward_fill"

        # Backward fill
        if after:
            interpolated["close"] = after[0]["close"]
            interpolated["open"] = after[0].get("open", after[0]["close"])
            interpolated["high"] = after[0].get("high", after[0]["close"])
            interpolated["low"] = after[0].get("low", after[0]["close"])
            return interpolated, "backward_fill"

        self._ensure_ohlc_consistency(interpolated)
        return interpolated, "no_interpolation"

    def _get_typical_high_close_ratio(self, before: list[dict], after: list[dict]) -> float:
        """Get typical High/Close ratio from surrounding prices."""
        ratios = []

        if before and before[0].get("close", 0) > 0:
            ratios.append(before[0].get("high", before[0]["close"]) / before[0]["close"])
        if after and after[0].get("close", 0) > 0:
            ratios.append(after[0].get("high", after[0]["close"]) / after[0]["close"])

        if not ratios:
            return 1.02  # Default: High is 2% above Close

        return sum(ratios) / len(ratios)

    def _get_typical_low_close_ratio(self, before: list[dict], after: list[dict]) -> float:
        """Get typical Low/Close ratio from surrounding prices."""
        ratios = []

        if before and before[0].get("close", 0) > 0:
            ratios.append(before[0].get("low", before[0]["close"]) / before[0]["close"])
        if after and after[0].get("close", 0) > 0:
            ratios.append(after[0].get("low", after[0]["close"]) / after[0]["close"])

        if not ratios:
            return 0.98  # Default: Low is 2% below Close

        return sum(ratios) / len(ratios)

    def _ensure_ohlc_consistency(self, price: dict) -> None:
        """Ensure OHLC consistency."""
        close = price.get("close", 0)
        open_price = price.get("open", close)
        high = price.get("high", close)
        low = price.get("low", close)

        # Ensure High >= all
        price["high"] = max(high, open_price, close)

        # Ensure Low <= all
        price["low"] = min(low, open_price, close)

        # Ensure High >= Low
        if price["high"] < price["low"]:
            price["high"] = price["low"]

    def validate_and_interpolate(self, prices: list[dict]) -> list[dict]:
        """
        Validate all prices and interpolate abnormal ones.
        Input prices should be in chronological order (oldest first).
        Returns validated price list.
        """
        if not prices:
            return prices

        result = []
        interpolation_count = 0

        for i, price in enumerate(prices):
            # Get previous price from result (already validated prices)
            previous_price = result[-1] if result else None

            # Build context from already-validated prices (last 30 days)
            context = list(reversed(result[-CONTEXT_WINDOW_DAYS:])) if result else []

            validation = self.validate_price(price, previous_price, context)

            if validation.all_valid():
                result.append(price)
                continue

            # Find before/after prices for interpolation
            before, after = self._find_interpolation_context(i, prices, result)

            interpolated, method = self.interpolate_price(price, validation, before, after)
            interpolation_count += 1

            result.append(interpolated)

        # Log warning if too many prices were interpolated
        if prices and interpolation_count / len(prices) > 0.5:
            logger.warning(f"More than 50% of prices flagged invalid ({interpolation_count}/{len(prices)})")

        return result

    def validate_price_series_desc(self, prices: list[dict]) -> list[dict]:
        """Validate a price series in descending order (newest first).

        Handles the reverse-validate-reverse pattern: reverses to chronological order,
        validates, and returns in the original descending order.

        Args:
            prices: Price list in descending order (newest first)

        Returns:
            Validated price list in descending order
        """
        if not prices:
            return prices
        return list(reversed(self.validate_and_interpolate(list(reversed(prices)))))

    def _find_interpolation_context(
        self, index: int, prices: list[dict], result: list[dict]
    ) -> tuple[list[dict], list[dict]]:
        """Find valid before/after prices for interpolation."""
        before = []
        after = []

        # Look for "before" price in validated results
        if result:
            before = [result[-1]]

        # Look for "after" price in remaining prices
        for j in range(index + 1, len(prices)):
            next_validation = self.validate_price(prices[j], None, result)
            if next_validation.all_valid():
                after = [prices[j]]
                break

        return before, after


def check_trade_blocking(current_price: float, historical_prices: list[float], symbol: str = "") -> Tuple[bool, str]:
    """
    Check if trades should be blocked due to price anomaly.

    Ported from sentinel2's scoring.CheckTradeBlocking.

    Logic:
    - High anomaly (price > 10x avg): BLOCK TRADES (prevents buying at inflated API error prices)
    - Low anomaly (price < 0.1x avg): ALLOW TRADES (may be genuine crash/opportunity)
    - No anomaly: ALLOW TRADES

    Args:
        current_price: The current live price from the API
        historical_prices: List of recent closing prices (last 30+ days)
        symbol: Symbol for logging (optional)

    Returns:
        Tuple of (allow_trade, reason)
        - allow_trade: True if trades are allowed
        - reason: Explanation if trades are blocked, empty string if allowed
    """
    if len(historical_prices) < 30:
        # Insufficient data for anomaly detection - allow trades
        return True, ""

    if current_price <= 0:
        return False, "invalid current price (zero or negative)"

    # Use last 30 days for context (excluding current price which isn't in historical)
    context_prices = historical_prices[-30:]

    if not context_prices:
        return True, ""

    avg_price = sum(context_prices) / len(context_prices)

    if avg_price <= 0:
        return True, ""

    ratio = current_price / avg_price

    # BLOCK on high anomaly only (likely API error)
    if ratio >= MAX_PRICE_MULTIPLIER:
        return False, f"high price anomaly: current {current_price:.2f} is {ratio:.1f}x the 30-day avg {avg_price:.2f}"

    # ALLOW on low anomaly (may be genuine crash/opportunity)
    # ALLOW on normal prices
    return True, ""


def get_price_anomaly_warning(current_price: float, historical_prices: list[float], symbol: str = "") -> Optional[str]:
    """
    Get a warning message if the current price appears anomalous.

    Unlike check_trade_blocking, this returns warnings for BOTH high and low anomalies
    so users can see when something unusual is happening.

    Returns:
        Warning message string, or None if price appears normal
    """
    if len(historical_prices) < 30:
        return None

    if current_price <= 0:
        return "Current price is zero or negative"

    context_prices = historical_prices[-30:]

    if not context_prices:
        return None

    avg_price = sum(context_prices) / len(context_prices)

    if avg_price <= 0:
        return None

    ratio = current_price / avg_price

    if ratio >= MAX_PRICE_MULTIPLIER:
        return (
            f"Price spike detected: {current_price:.2f} is {ratio:.1f}x the 30-day average"
            f" ({avg_price:.2f}). Trades blocked."
        )

    if ratio <= MIN_PRICE_MULTIPLIER:
        return (
            f"Price crash detected: {current_price:.2f} is {ratio:.1f}x the 30-day average"
            f" ({avg_price:.2f}). Verify before trading."
        )

    return None
